tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,644 @@
+                    self.io.registration().readiness(Interest::WRITABLE).await?;
+                }
+                Err(e) => return Err(e),
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Disconnects the server end of a named pipe instance from a client
+    /// process.
+    ///
+    /// ```no_run
+    /// use tokio::net::windows::NamedPipeBuilder;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    /// let pipe = builder.create()?;
+    ///
+    /// // Wait for a client to connect.
+    /// pipe.connect().await?;
+    ///
+    /// // Use the pipe...
+    ///
+    /// // Forcibly disconnect the client (optional).
+    /// pipe.disconnect()?;
+    /// # Ok(()) }
+    /// ```
+    pub fn disconnect(&self) -> io::Result<()> {
+        self.io.disconnect()
+    }
+}
+
+impl AsyncRead for NamedPipe {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        unsafe { self.io.poll_read(cx, buf) }
+    }
+}
+
+impl AsyncWrite for NamedPipe {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        self.io.poll_write(cx, buf)
+    }
+
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[io::IoSlice<'_>],
+    ) -> Poll<io::Result<usize>> {
+        self.io.poll_write_vectored(cx, bufs)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        self.poll_flush(cx)
+    }
+}
+
+/// Raw handle conversion for [NamedPipe].
+///
+/// # Panics
+///
+/// This panics if called outside of a [Tokio Runtime] which doesn't have [I/O
+/// enabled].
+///
+/// [Tokio Runtime]: crate::runtime::Runtime
+/// [I/O enabled]: crate::runtime::RuntimeBuilder::enable_io
+impl FromRawHandle for NamedPipe {
+    unsafe fn from_raw_handle(handle: RawHandle) -> Self {
+        Self::try_from_raw_handle(handle).unwrap()
+    }
+}
+
+impl AsRawHandle for NamedPipe {
+    fn as_raw_handle(&self) -> RawHandle {
+        self.io.as_raw_handle()
+    }
+}
+
+// Helper to set a boolean flag as a bitfield.
+macro_rules! bool_flag {
+    ($f:expr, $t:expr, $flag:expr) => {{
+        let current = $f;
+
+        if $t {
+            $f = current | $flag;
+        } else {
+            $f = current & !$flag;
+        };
+    }};
+}
+
+/// A builder structure for construct a named pipe with named pipe-specific
+/// options. This is required to use for named pipe servers who wants to modify
+/// pipe-related options.
+///
+/// See [NamedPipeBuilder::create].
+#[derive(Clone)]
+pub struct NamedPipeBuilder {
+    name: Arc<[u16]>,
```

Comment:
Why is this an `Arc`?

Reply:
Cheap to clone which a listening server needs to that a bit and implements `Sync` and `Send`. Can move it back to a `Box<[u16]>` though if wanted.

Reply:
Your build methods don't consume the builder, so I don't think using an `Arc` for that is necessary.

Reply:
Yeah, I was considering it initially. So which one do you prefer `OwnedPipeName` / `PipeName` combo or just `&PipeName`?

Reply:
I think it would be nice if we choose some solution where people who don't care can pass an `&str`.

Reply:
For now I'll just change it back to a `Box<[u16]>`. But at least now people can see my broader thinking in text.

Reply:
Modified everything to use `*Options` structures and we no longer store the names.